pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -101,22 +103,65 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
-			for _, n := range watched.ResourceNames {
-				_, _, svcHost, _ := model.ParseSubsetKey(n)
-				if svcHost == host.Name(key.Name) {
-					deletedClusters = append(deletedClusters, n)
-				}
+		// get service that changed
+		if key.Kind == gvk.ServiceEntry {
+			service := updates.Push.ServiceForHostname(proxy, host.Name(key.Name))
+			// SidecarScope.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+			// we can cross-reference with WatchedResources to figure out which services were deleted.
+			if service == nil {
+				// we assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+				// does not exist.
+				// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+				// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+				// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
+				deletedClusters = append(deletedClusters, getRemovedClusterNames(watched.ResourceNames, nil, key.Name)...)
+			} else {
+				services = append(services, service)
 			}
 		} else {
-			services = append(services, service)
+			// destination rules
+			cfg := updates.Push.DestinationRuleByName(proxy, key.Name, key.Namespace)
+			log.Infof("adiprerepa: destinationRule name: %v namespace: %v", key.Name, key.Namespace)
+			if cfg == nil {
+				// subset clusters need to be removed, actual cluster needs to be updated
+				// a destinationrule was deleted, get the populated destinationRule from PrevSidecarScope
+				prevCfg := updates.Push.PrevDestinationRuleByName(proxy, key.Name, key.Namespace)
+				if prevCfg == nil {
+					// something went wrong, skip -- we shouldn't be nil here
+					break
+				}
+				// populate removed clusters ONLY for subsets.
+				dr := prevCfg.Spec.(*networking.DestinationRule)
+				deletedClusters = append(deletedClusters, getRemovedClusterNames(subsetClusters(watched.ResourceNames), nil, dr.Host)...)
+
+				// for non-subset destinationrule deletion -- generate the cluster as it changed
+				matched := updates.Push.ServicesForHostname(proxy, host.Name(dr.Host))
+				services = append(services, matched...)
+			} else {
+				// destination exists, was updated
+				// generate cluster based on service associated with destinationrule
+				// hostName can be a wildcard, and can match to multiple services
+				dr := cfg.Spec.(*networking.DestinationRule)
+				matchedSvcs := updates.Push.ServicesForHostname(proxy, host.Name(dr.Host))
+				services = append(services, matchedSvcs...)
+				// check for removed subsets
+				prevCfg := updates.Push.PrevDestinationRuleByName(proxy, key.Name, key.Namespace	)
+				if prevCfg == nil {
+					log.Infof("Prev DestinationRule form PrevSidecarScope is nil for %v", key.String())
+					break
```

Comment:
is break correct? Should n't we continue with other configs?

Reply:
correct, fixed